### ISSUE TITLE
TypeScript 4.6 regression

### ISSUE THREAD

**Issue by thetutlage** [microsoft/TypeScript#48118]
# Bug Report

I have read the release notes and the code that is breaking for me is not mentioned in the breaking changes section.

### 🔎 Search Terms
None. As the code works with 4.5

### 🕗 Version & Regression Information
The `typeof var === 'function' call narrows the variable to function vs one of the union.

- This changed between versions `4.5` and 4.6

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.2#code/C4TwDgpgBAwg9gO2BAHsKBeKAlCBjOAJwBMAeAZ2EIEsEBzAGigEMEQA+AWACge8AbZuXJQAYnDikAKlFTIExEfCRz2UAN5QoPLWBoA3ZsigEVaAIJ48AVwC2g4EQCEALigAKAJSY1MgD5QAAqEcLbU5BDSXLzcWqaUhNZ4joTuOlomiMhobv4eXj5QecGh4ZFS7J7p3urpWtQAZh6gkHBNptnoGN1QAOQN1gjJ1Ii9NXUZwAAW4QB0HXKWNvZGRJiZZsATAL7pu9zbQA)

### 💻 Code

```ts
type Context = Record<string, any>

class Foo<T extends Context> {  
  private contextAccumlator!: () => T | Promise<T>

  constructor(
    context: T | (() => T | Promise<T>)
  ) {
    if (typeof context === 'function') {
      this.contextAccumlator = context
    }
  }
}
```

### 🙁 Actual behavior

I expect `this.contextAccumlator = context` to report zero errors.

### 🙂 Expected behavior

Instead it says

```
Type '(() => T | Promise<T>) | (T & Function)' is not assignable to type '() => T | Promise<T>'.
  Type 'T & Function' is not assignable to type '() => T | Promise<T>'.
    Type 'Context & Function' provides no match for the signature '(): T | Promise<T>'.(2322)
```

**Comment by tristan00b** [comment]
@thetutlage I played around with this a bit and reduced your example:

```ts
function isFunction<T extends Function>(it: unknown): it is T
{
  return typeof it === 'function'
}

function isFn(it: unknown): it is Function
{
  return typeof it === 'function'
}

function f<T>(arg: T | (() => T))
{
  let x: () => T
 
  if (typeof arg === 'function')
    x = arg // (error) arg: (() => T) | (T & Function)

  if (isFunction(arg))
    x = arg // (works) arg: () => T

  if (isFn(arg))
    x = arg // (works) arg: () => T
}
```

This can be reproduced as far back as version 3.3.3 ([play](https://www.typescriptlang.org/play?target=8&jsx=0&ts=4.6.2#code/GYVwdgxgLglg9mABDAzgMXNeYA8AVRAUwA8pCwATFRDSWBAPgAoYoAuRcAazDgHcwASg6tk1PACgA3hMSIAToSgh5SKAE8ADoTjBkURAF5jiAOSg62UxIC+EiRawIxaMC3acwPfkJEHUNJj0YNKyCkoqalo6eqLGhmaOwdZ2DkHYiMD4zACG8gDmHAQAPohMTIJGDIh4goKhcgA2SojEHBVVNbJhMHpMGtq6iHn5Ribm6Qim9XJyxEbDBYgA9MtlhPLycPKVI+0dhtW1iKVMBABkgZYI9T19qLRObiN1YXMLIytrTHzbXCi7ArtSqHLp3MoPZ4FV6zVofJarMq-eT-QGFMogo62IA)).

**Comment by thetutlage** [comment]
🤔 Did you mean there is no regression?

**Comment by jihndai** [comment]
Weirdly enough, it seems those are two different cases. This issue's sample code was indeed working by 4.5.5, broken in 4.6.2, and still broken in 4.7.0-dev.20220304. While the second example was never fixed as far as I can tell.

**Comment by tristan00b** [comment]
Maybe a recent change caused the old bug to surface in @thetutlage 's code.

**Comment by RyanCavanaugh** [comment]
This is a correct error that was not properly found before:
```ts
type Context = Record<string, any>

class Foo<T extends Context> {  
  private contextAccumlator!: () => (T | Promise<T>)

  constructor(context: T | (() => T | Promise<T>)
  ) {
    if (typeof context === 'function') {
      this.contextAccumlator = context
    }
  }

  invoke() {
    this.contextAccumlator();
  }
}

function isPromise(arg: any): arg is Promise<any> {
  return false;
}

const p = new Foo<(n: string) => void>(n => n.substring(0));
// Crashes
p.invoke();
```

**Comment by thetutlage** [comment]
So shouldn't this be under breaking change?

**Comment by RyanCavanaugh** [comment]
Yes, that's why it's a Docs bug assigned to the blog post author

**Comment by thetutlage** [comment]
Okay! Thanks for the clarification :)

**Comment by jakebailey** [comment]
This was "fixed" in #49919; should it have been?
